### PR TITLE
s3-cache-pull 0.1.0

### PR DESCRIPTION
**Pull request by alephao**
TagCheck

### What to do if the build fails?

At the moment contributors do not have access to the CI workflow triggered by StepLib PRs. In case of a failed build, we ask for your patience. Maintainers of Bitrise Steplib will sort it out for you or inform you if any further action is needed.

### New Pull Request Checklist

*Please mark the points which you did / accept.*

- [x] __I will not move an already shared step version's tag to another commit__
- [x] I read the [Step Development Guideline](https://github.com/bitrise-io/bitrise/blob/master/_docs/step-development-guideline.md)
- [x] I have a test for my Step, which can be run with `bitrise run test` (in the step's repository)
- [x] I did run `bitrise run audit-this-step` (in the step's repository - note, if you don't have this workflow in your `bitrise.yml`, [you can copy it from the step template](https://github.com/bitrise-steplib/step-template/blob/master/bitrise.yml).)
- [x] I read and accept the [Abandoned Step policy](https://github.com/bitrise-io/bitrise-steplib#abandoned-step-policy)

**New Step**
Thank you for the new Step share! The CI check might will fail due to our extended validation engine. Nothing to worry about yet, we will get back to you shortly.

### CODE REVIEW

**Review by trapacska — APPROVED** [review]

**Comment by mateherber** [comment]
Hi @alephao,

Thank you for your contribution. I'm sorry we haven't had the resources to start the review earlier.

I'd like to have a couple of questions regarding the implementation. 
1) The pull & push steps share a significant amount of code in the parser package. Do you plan to add it as a module dependency, or a git submodule?
2) Both the pull and push step is used to handle 1 archive right? So if I wanted to push/pull multiple files/dirs I would either have to move them in one directory or add the pull and push steps multiple times in my workflow.
3) The cache push works in a way if it finds a key present in S3 it doesn't upload new content. So essentially once a file is written with a key to the storage it can never change, right?

For the sake of science, I'd like to ask whether you checked out our cache offering ([cache-push](https://www.bitrise.io/integrations/steps/cache-push) and [cache-pull](https://www.bitrise.io/integrations/steps/cache-pull)) steps. It does a similar thing to these steps however it handles the key generation a bit differently (and uploads the entire content of the cache in one archive).

Apart from the fact that it doesn't upload to S3 is there anything in those steps' mechanism that doesn't fit your use-case? We'd like to know how could we improve our caching so there is no need for multiple cache steps in the step library in the long run.

File: steps/s3-cache-pull/0.1.0/step.yml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  commit: 98a2b4001302f3d1e9309f57cdb788fb3b3d19b4
+host_os_tags:
+- osx-10.10
+- ubuntu-16.04
+type_tags:
+- utility
+toolkit:
+  go:
+    package_name: github.com/alephao/bitrise-step-s3-cache-pull
+deps:
```
Comment (mateherber) [inline steps/s3-cache-pull/0.1.0/step.yml:19]: Please remove the unused dependencies

File: steps/s3-cache-pull/0.1.0/step.yml [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  - name: wget
+  apt_get:
+  - name: git
+  - name: wget
+is_requires_admin_user: true
+is_always_run: false
+is_skippable: false
+run_if: ""
+inputs:
+- aws_access_key_id: $AWS_ACCESS_KEY_ID
```
Comment (mateherber) [inline steps/s3-cache-pull/0.1.0/step.yml:31]: Please don't use default env var values when it is not expected to be available. (exported by other steps as outputs or prefilled by bitrise webhook handling). This applies to the rest of the inputs too.

File: steps/s3-cache-pull/0.1.0/step.yml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  opts:
+    category: AWS Access
+    is_expand: true
+    is_required: false
```
Comment (mateherber) [inline steps/s3-cache-pull/0.1.0/step.yml:35]: This should be required

File: steps/s3-cache-pull/0.1.0/step.yml [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    title: AWS_ACCESS_KEY_ID
+- aws_secret_access_key: $AWS_SECRET_ACCESS_KEY
+  opts:
+    category: AWS Access
+    is_expand: true
+    is_required: false
```
Comment (mateherber) [inline steps/s3-cache-pull/0.1.0/step.yml:41]: This should be required

File: steps/s3-cache-pull/0.1.0/step.yml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  apt_get:
+  - name: git
+  - name: wget
+is_requires_admin_user: true
+is_always_run: false
+is_skippable: false
+run_if: ""
+inputs:
+- aws_access_key_id: $AWS_ACCESS_KEY_ID
+  opts:
```
Comment (mateherber) [inline steps/s3-cache-pull/0.1.0/step.yml:32]: Please add `summary` and `description` to **all** inputs

File: steps/s3-cache-pull/0.1.0/step.yml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    category: AWS Access
+    is_expand: true
+    is_required: false
+    title: AWS_ACCESS_KEY_ID
+- aws_secret_access_key: $AWS_SECRET_ACCESS_KEY
+  opts:
```
Comment (mateherber) [inline steps/s3-cache-pull/0.1.0/step.yml:38]: please use the `is_sensitive: true` option on **all** inputs that are sensitive

File: steps/s3-cache-pull/0.1.0/step.yml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    - me-south-1
+    - af-south-1
+- bucket_name: $S3_BUCKET_NAME
+  opts:
+    category: AWS Bucket
+    is_expand: true
+    is_required: true
+    summary: The bucket name where you cache is stored
+    title: Bucket Name
+- opts:
```
Comment (mateherber) [inline steps/s3-cache-pull/0.1.0/step.yml:85]: Names of the inputs are off here, should be `restore_keys` and should contain `opts`

File: steps/s3-cache-pull/0.1.0/step.yml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    category: Cache
+    description: "You can use '{{ checksum path/to/file }}' to get the file's sha256
+      checksum.      \nYou can use '{{ branch }}' to get the name of the current branch.\nE.g.:\nrestore_keys:
+      |\n  carthage-{{ branch }}-{{ checksum \"Cartfile.resolved\" }}\n  carthage-\n"
+    is_expand: false
+    is_required: true
+    summary: List of keys with fallbacks to restore the cache, one per line
+    title: Cache restore keys
+  restore_keys: null
```
Comment (mateherber) [inline steps/s3-cache-pull/0.1.0/step.yml:94]: `null` can be omitted here

File: steps/s3-cache-pull/0.1.0/step.yml [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+- opts:
```
Comment (mateherber) [inline steps/s3-cache-pull/0.1.0/step.yml:95]: Name of the input should be `path` and contain `opts`

**Comment by mateherber** [comment]
Hi @alephao,

Thanks for the detailed answer I'll bring these back to the team.  I've reviewed the step.yml and there are some changes we'd like to request. Please fix these in the `step.yml` of the step and reopen a steplib PR. I'm going to close this PR. (Unfortunately, this is the easiest way to update)

**Pull request closed without merging**